### PR TITLE
implement past course run display

### LINKED ISSUE #3049: Dashboard: Completed grades details view
Once a user has completed the edX course and proctored exam, we will show their grades in the course card (see #2792). The grades link to a modal with a history of ALL grades in the course or exam.

When the modal is displayed, gray out the entire page with a transparent white overlay (as shown). Clicking "x" or anywhere outside the modal, closes it.

Any passing grade is highlighted in green with the word "Passed". The highest passing grade displays a checkmark.

image

### PR DESCRIPTION
**Pull request by aliceriot**
#### What are the relevant tickets?

this closes #3049 

#### What's this PR do?

this implements one part that was left off of the dashboard redesign PR to make it a little more manageable, the display of past course runs.

#### How should this be manually tested?

Basically, load up a bunch of different course runs and make sure the display of past runs makes sense!

You need to click on the `<Grades />` component to get the course run popup to show itself.

### CODE REVIEW

**Comment by noisecapella** [comment]
Can you make the cursor a hand when you hover over the `<Grade />` component?

File: static/js/components/dashboard/courses/GradeDetailPopup.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  Course,
+  CourseRun,
+} from '../../../flow/programTypes';
+import { formatGrade } from '../util';
+import {
+  STATUS_NOT_PASSED,
+  STATUS_PASSED,
+} from '../../../constants';
+
+const runGrade = (courseRun: CourseRun, isBestGrade) => {
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: Can you add a type for `isBestGrade` and the return type?

File: static/js/components/dashboard/courses/GradeDetailPopup.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  return "";
+};
+
+const runStatus = courseRun => {
+  if (courseRun.status === STATUS_NOT_PASSED) {
+    return <div className="status failed">
+      Not passed
+    </div>;
+  }
+  if (courseRun.status === STATUS_PASSED || courseRun.passed === true) {
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: I don't think `passed` is a property of `courseRun`. Also can you add a flow annotation for this argument?

File: static/js/components/dashboard/courses/GradeDetailPopup.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    return <div className="status passed">
+      Passed
+    </div>;
+  }
+  return <div className="status audited">
+    Audited
+  </div>;
+};
+
+const renderRunRow = ([courseRun, isBestGrade], idx) => (
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: Can you add flow annotations here too?

File: static/js/components/dashboard/courses/GradeDetailPopup.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  <div className="course-run-row" key={idx}>
+    <div className="title">{ courseRun.title }</div>
+    <div className="grade-status">
+      <div>{ runStatus(courseRun) }</div>
+      <div className="grade">{ runGrade(courseRun, isBestGrade) }</div>
+    </div>
+  </div>
+);
+
+const labelBestGrade = runs => {
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: Can you add a flow annotation here?

File: static/js/components/dashboard/courses/GradeDetailPopup.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    [run, run.final_grade && run.final_grade !== 0 && run.final_grade === bestGrade]
+  ));
+};
+
+const renderRunRows = R.compose(
+  R.addIndex(R.map)(renderRunRow),
+  labelBestGrade
+);
+
+const dialogTitle = course => (
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: Can you add a flow annotation here?

**Comment by noisecapella** [comment]
Code and tests look good overall, just some nitpicking

File: static/js/components/dashboard/courses/GradeDetailPopup.js (hunk 2 above)
Comment (aliceriot) [inline static/js/components/dashboard/courses/GradeDetailPopup.js]: oh haha I was confusing that with the exam grades :P

**Comment by aliceriot** [comment]
comments addressed!

**Comment by noisecapella** [comment]
Can you address the other two comments regarding adding tests? I 'requested changes' twice so it wasn't grouped together with the other comments

**Comment by aliceriot** [comment]
I added a test that should exercise that function - what were you suggesting specifically?

File: static/js/components/dashboard/courses/Grades_test.js [hunk 6]
```diff
@@ -123,4 +129,10 @@ describe('Course Grades', () => {
     course.has_exam = false;
     assert.equal(renderGrades().find('.passed-course').length, 0);
   });
+
+  it('should call setShowGradeDetailDialog onClick', () => {
```
Comment (noisecapella) [inline static/js/components/dashboard/courses/Grades_test.js]: Can you also test that if you click it when it's visible it calls `setShowGradeDetailDialog` with `false?

Comment (aliceriot) [inline static/js/components/dashboard/courses/Grades_test.js]: I added a test that calls the `onRequestClose` prop on the dialog - I had trouble getting a click or keypress based test to work correctly.

File: static/js/containers/LearnerPage_test.js [hunk 7]
```diff
@@ -1008,6 +1013,18 @@ describe("LearnerPage", function() {
       });
     });
 
+    it('should show grades, if the user has a staff role', () => {
```
Comment (noisecapella) [inline static/js/containers/LearnerPage_test.js]: Can you also add a test for hiding the dialog, similar to the one for `DashboardPage`?

**Comment by aliceriot** [comment]
@noisecapella I added a test!

**Pull request merged**